staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go

```diff
@@ -79,47 +79,41 @@ func NewCommandStartWardleServer(out, errOut io.Writer, stopCh <-chan struct{})
 
 	flags := cmd.Flags()
 	o.RecommendedOptions.AddFlags(flags)
-	o.Admission.AddFlags(flags)
 
 	return cmd
 }
 
 func (o WardleServerOptions) Validate(args []string) error {
 	errors := []error{}
 	errors = append(errors, o.RecommendedOptions.Validate()...)
-	errors = append(errors, o.Admission.Validate()...)
 	return utilerrors.NewAggregate(errors)
 }
 
 func (o *WardleServerOptions) Complete() error {
 	return nil
 }
 
-func (o WardleServerOptions) Config() (*apiserver.Config, error) {
+func (o *WardleServerOptions) Config() (*apiserver.Config, error) {
 	// register admission plugins
-	banflunder.Register(o.Admission.Plugins)
+	banflunder.Register(o.RecommendedOptions.Admission.Plugins)
 
 	// TODO have a "real" external address
 	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts("localhost", nil, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
 		return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
 	}
 
 	serverConfig := genericapiserver.NewRecommendedConfig(apiserver.Codecs)
-	if err := o.RecommendedOptions.ApplyTo(serverConfig); err != nil {
-		return nil, err
-	}
 
-	client, err := clientset.NewForConfig(serverConfig.LoopbackClientConfig)
-	if err != nil {
-		return nil, err
+	o.RecommendedOptions.ExtraAdmissionInitializers = func() ([]admission.PluginInitializer, error) {
+		client, err := clientset.NewForConfig(serverConfig.LoopbackClientConfig)
+		if err != nil {
+			return nil, err
+		}
+		informerFactory := informers.NewSharedInformerFactory(client, serverConfig.LoopbackClientConfig.Timeout)
+		o.SharedInformerFactory = informerFactory
+		return []admission.PluginInitializer{wardleinitializer.New(informerFactory)}, nil
 	}
-	informerFactory := informers.NewSharedInformerFactory(client, serverConfig.LoopbackClientConfig.Timeout)
-	admissionInitializer, err := wardleinitializer.New(informerFactory)
-	if err != nil {
-		return nil, err
-	}
-
-	if err := o.Admission.ApplyTo(&serverConfig.Config, serverConfig.SharedInformerFactory, serverConfig.ClientConfig, apiserver.Scheme, admissionInitializer); err != nil {
+	if err := o.RecommendedOptions.ApplyTo(serverConfig, apiserver.Scheme); err != nil {
```

Comment:
@deads2k please take a look here. If you do nothing (and there is no need for extra initializers), the admission is initialized (which can include webhooks, probably not wired yet). If you need extra initializers, `ExtraAdmissionInitializers` is a hook to inject them. It's using a closure because we depend on `CoreAPI.ApplyTo` to be called before to get a loopback client config. I still have the feeling it is not super elegant, but maybe good enough.

Reply:
Yeah, it definitely lacks elegance.  My mind would be somewhat eased if you passed the serverConfig in as an arg.